### ISSUE TITLE
New permissions record not being ingested

### ISSUE THREAD

**Issue by JosephMcArthur** [oaworks/discussion#1765]
[`https://dev.lvatn.com/service/oab/permissions/Thomas%20Telford%20Ltd.`](https://dev.lvatn.com/service/oab/permissions/Thomas%20Telford%20Ltd.) should have two records, as it does in the database, but they don't seem to be being ingested.

The records can be found on lines [228 and 229 of the ingest sheet](https://docs.google.com/spreadsheets/d/1qBb0RV1XgO3xOQMdHJBAf3HCJlUgsXqDVauWAtxde4A/edit#gid=116570718).

This is for a Plan S thing, so would be great to address quickly.

**Comment by markmacgillivray** [comment]
I haven't looked at this yet.

**Comment by markmacgillivray** [comment]
Looks to me like it works fine though. For a while the sheet import was just finding an empty main sheet again, I assumed you just hadn't populated it again for some reason. Since that often seems to happen, I just have it do nothing if the sheet returns nothing. Then later once there are records again, it imports them. I guess this can be closed unless I'm missing something.

**Comment by markmacgillivray** [comment]
Now there is only one but still two in the sheet. Will have to find out what is causing it to not always ingest them.

**Comment by JosephMcArthur** [comment]
A quick reminder... the dev repo is here: https://github.com/JosephMcArthur/rscvd-dev

**Comment by JosephMcArthur** [comment]
> A quick reminder... the dev repo is here: https://github.com/JosephMcArthur/rscvd-dev

Whooops, wrong issue...

**Comment by JosephMcArthur** [comment]
Folks are also asking us about this.

**Comment by markmacgillivray** [comment]
I've been testing this and as long as everything is in the sheet it appears to import them all. I've done multiple imports and each time I see all of them appear now. That's 10488 at the moment. And I see the two mentioned above, and they both get returned in the permissions check. So I think this can be closed unless you are still seeing something different (on dev).

**Comment by JosephMcArthur** [comment]
Did you make any changes to have this be the case?

**Comment by markmacgillivray** [comment]
No. But it just ran itself automatically on schedule and the email alert tells me it's back to 10464. So I'll have to have another look.

**Comment by markmacgillivray** [comment]
I have checked now and the automatic one was the live one. It did not import the two records, and is showing less than dev. So I suspect it is just that dev has been updated with the necessary changes to allow these new records but we have not yet pushed that to live. Checking again on dev just now I can see the 10488 getting imported, so it is definitely importing all of them on dev.

**Comment by markmacgillivray** [comment]
The auto scheduled load for dev just ran and it got 10488 as well, so definitely working on dev even when on auto, not me manually running and checking.

**Comment by JosephMcArthur** [comment]
Can I get this confirmation, and can we just keep an eye on it for a while to see if the we see the issue arise again?

Joseph McArthur
Assistant Director: Right to Research Coalition ( http://righttoresearch.org )
Director / Founder: Open Access Button ( http://openaccessbutton.org )
Twitter: @Mcarthur_Joe ( https://twitter.com/Mcarthur_Joe )
Skype: joseph_mcarthur

**Comment by markmacgillivray** [comment]
Yep done

**Comment by markmacgillivray** [comment]
This has come back again and I've been monitoring it. It's not our system. Direct checks on what we get from google indicate that we're getting different versions from the google request. I tried adding random params in case if intermediate caching but that had no effect. No other solution in mind yet, apart from doing the full google API work that we haven't scheduled yet. Will keep watching it and looking for alternative solutions.

**Comment by JosephMcArthur** [comment]
I think I have a hunch as to the problem. Hopefully I"ll be able to take a look today.

Joseph McArthur
Assistant Director: Right to Research Coalition ( http://righttoresearch.org )
Director / Founder: Open Access Button ( http://openaccessbutton.org )
Twitter: @Mcarthur_Joe ( https://twitter.com/Mcarthur_Joe )
Skype: joseph_mcarthur

**Comment by JosephMcArthur** [comment]
Dumping some info here: 

> Just to confirm more, over the last few days I can see some imports getting 10482 and 10483 records, which I would guess are the right amounts for the times, but then numerous times inbetween it falls back to 10464 records, which is the number it seems to sometimes get stuck on. So something in the sheet is causing the sheet to only export that old amount sometimes, for some unknown reason, and intermittently.

> still have issues with https://api.openaccessbutton.org/permissions/2053-2296

`OAB permissions import check complete` still reports fluctuating results sometimes

**Comment by JosephMcArthur** [comment]
@natalianorori and I worked on this today. Some notes are [here](https://docs.google.com/document/d/1Nh8HXy1X1CuOQ0_qWgET1JlPk_NTGgDiRZqHuXOQsRU/edit#).

@markmacgillivray, we need you to make a change.

Instead of importing all permissions data from [here](https://docs.google.com/spreadsheets/d/1qBb0RV1XgO3xOQMdHJBAf3HCJlUgsXqDVauWAtxde4A/edit#gid=1575040992) we'd like to have the system import each data type separately. This should reduce the chance of inconsistencies, and if there are still issues, help us in tracking them down.

The new sheets to import permissions data from are:

* [Journals](https://docs.google.com/spreadsheets/d/19pDvOY5pge-C0yDSObnkMqqlMJgct3iIjPI2rMPLQEc/edit#gid=1540153499)
* [Publishers](https://docs.google.com/spreadsheets/d/1tmEfeJ6RCTCQjcCht-FI7FH-04z7MPSKdUnm0UpAxWM/edit#gid=2103194526 )
* [Affiliations](https://docs.google.com/spreadsheets/d/1J4WhZjPsAjpoogsj7wSTQGJPguo7TiSe0uNcrvyd_OM/edit#gid=1810976998)

Together that data will add up to what used to be exported.

We think that this, along with the other changes we're making, should allow us to solve this. However, we'll monitor that.

**Comment by JosephMcArthur** [comment]
Also, @markmacgillivray, if you could make this change early in the week that would be ideal as Natalia and I are working on #1590 together atm and this will impact that work.

**Comment by markmacgillivray** [comment]
Hope you don't mind if I ask some questions: 

why do you import large amounts of data from one sheet to another? Wouldn't it be easier to just have the API import from the original sheet? Some of the imports don't seem to make any changes to the data. There are limitations to what we can do with the simple "publish-import", but we can target a specific sheet as long you don't move that sheet later.

At least one of the new sheets you linked to has an empty row in it - that will break the publish-import, on the google side it stops including records in the publish once it hits an empty row.

If you're going to refactor the sheets anyway, would it be possible to start using simplified column headers? If they're lowercased, don't have special characters, use underscores instead of whitespace, it's always going to keep things easier and less likely to break when automating stuff. Also, we could match up the names used in the sheet with the names used on the API, instead of it having to translate.

Timestamps are not strictly necessary for more regular updates of the data. We can run an import any time, but doing so requires us to reload all data because they don't have reliable IDs. With a timestamp that you update whenever you make a change, we could only import records that have changed. However, at the scale of records this is at, it's easy for the system to import the whole lot again anyway. Also, if you think it would be possible not to change the order of records, then their row number could be used as an ID instead. However, we probably don't need either of these solutions now because I've already put work into trying to make it use the values we currently have in there as IDs, plus see my next comment.

The reason it was inefficient to import every 15 minutes was that if there isn't any change it would just be unnecessary load, increases chance of error and downtime, etc, and the occasional errors we were seeing before were indeed causing problems. But if @natalianorori is going to trigger an import to the API when she wants to ensure changes are in the system, that's actually already possible, and would be a more efficient way of doing it anyway - just hit a certain URL when you want an import, I can tell you what that is, and it will do an import. It also shows you the number it imported as soon as it's done, instead of waiting for an email.

We want to generalise how we can integrate the use of sheets in the new API so we can make more use of this approach, and do it more easily in future. We can make the import smarter by doing extra work on the API to use the full Google sheets API instead of just doing the publish-import mechanism we use now, which would reduce some of the fragility around things like blank rows breaking the import. However it would be a bit of work to start using google APIs for this and dealing with the auth requirements in the sheets. If you think using multiple sheets that should import to the same index is something that will help you manage sheets well now and in future, it's just a little bit of work for me to add that now with the publish-import mechanism. But note, that may make it harder if we do later want to start having the API be able to update the sheets directly - if there's more than one sheet per index, we'd have to keep track of which records came from which sheet. Or we could have an index per sheet, but then depending what you want the API to be able to do with combinations of the data, that could make things harder at that end.

For the immediate issue and what you've asked for above, I won't be able to do it unless the blank rows are removed from the target sheets. Also, I'd need to know if the column names you've used are the same. If they're not, it's quite a lot of work to write a custom translation for every sheet. Of course I'll do that, but it's unlikely I'd be able to get it done early in the week unless I do nothing else and this takes top priority. And I think if we're making changes to sheets anyway, it would make more sense to change how we name columns now, so that we can generalise the problem, rather than spending time writing translators for each sheet and any future sheet.

**Comment by markmacgillivray** [comment]
PS, things like differences of whitespace vs underscores in the column names is probably irrelevant because google changes those to lowercased underscored anyway, and strips some special characters. It makes it harder to know if it's going to error out, but we're already doing it that way anyway - what I mean about whether or not the column names are different is, have you used any actually different column names? If they only differ by whitespace/underscore or capitalisation, then we're probably OK for now.

**Comment by markmacgillivray** [comment]
Another thing: the one you've linked as the one we're currently using is called v3, whereas the others are v4 but appear to read from a sheet called v2.1. If v2.1 is the most up to date, is it possible to just read from the v2.1 sheets? What is the import to other sheets doing for you? And can we just avoid that or replicate it on the API side to make things easier and possibly less error-prone on the google-sheets side of things? I'd guess large imports of data between sheets is where we end up with errors when we try to import them from google.

**Comment by markmacgillivray** [comment]
Just wondering too, is the error of some records not loading still happening? I've not been explicitly checking, but recent emails seem to have been a more consistent number.

**Comment by JosephMcArthur** [comment]
> why do you import large amounts of data from one sheet to another? Wouldn't it be easier to just have the API import from the original sheet? Some of the imports don't seem to make any changes to the data. There are limitations to what we can do with the simple "publish-import", but we can target a specific sheet as long you don't move that sheet later.

The actual management system is a little messy and very large. The exports try to clean that up (e.g spliting up data, giving it correct headers, removing spaces).

Hoping to remove those issues soon, but for now we'd like to try and address this issue.

> At least one of the new sheets you linked to has an empty row in it - that will break the publish-import, on the google side it stops including records in the publish once it hits an empty row.

Ah, yes, sorry missed that was needed. Should be done now.

> If you're going to refactor the sheets anyway, would it be possible to start using simplified column headers? If they're lowercased, don't have special characters, use underscores instead of whitespace, it's always going to keep things easier and less likely to break when automating stuff. Also, we could match up the names used in the sheet with the names used on the API, instead of it having to translate.

Yes, will plan to do that when we refactor the main sheets.

> Timestamps are not strictly necessary for more regular updates of the data. We can run an import any time, but doing so requires us to reload all data because they don't have reliable IDs. With a timestamp that you update whenever you make a change, we could only import records that have changed. However, at the scale of records this is at, it's easy for the system to import the whole lot again anyway. Also, if you think it would be possible not to change the order of records, then their row number could be used as an ID instead. However, we probably don't need either of these solutions now because I've already put work into trying to make it use the values we currently have in there as IDs, plus see my next comment.
> The reason it was inefficient to import every 15 minutes was that if there isn't any change it would just be unnecessary load, increases chance of error and downtime, etc, and the occasional errors we were seeing before were indeed causing problems. But if @natalianorori is going to trigger an import to the API when she wants to ensure changes are in the system, that's actually already possible, and would be a more efficient way of doing it anyway - just hit a certain URL when you want an import, I can tell you what that is, and it will do an import. It also shows you the number it imported as soon as it's done, instead of waiting for an email.

Can you tell us what the url is and we'll use it.

> We want to generalise how we can integrate the use of sheets in the new API so we can make more use of this approach, and do it more easily in future. We can make the import smarter by doing extra work on the API to use the full Google sheets API instead of just doing the publish-import mechanism we use now, which would reduce some of the fragility around things like blank rows breaking the import. However it would be a bit of work to start using google APIs for this and dealing with the auth requirements in the sheets. If you think using multiple sheets that should import to the same index is something that will help you manage sheets well now and in future, it's just a little bit of work for me to add that now with the publish-import mechanism. But note, that may make it harder if we do later want to start having the API be able to update the sheets directly - if there's more than one sheet per index, we'd have to keep track of which records came from which sheet. Or we could have an index per sheet, but then depending what you want the API to be able to do with combinations of the data, that could make things harder at that end.

I'm having a hard time digesting this to make a decision. Maybe I'll say this... we don't intend our changes here to result in significant long term adjustments to how data is handled on your end or how we import it now. I am keeping in mind a discussion with you where we said that it would be clearer from an ID perspective to have each `type` of permission imported in it's own sheet. So, I also hope this isn't a step "backward". You and I may need to discuss if you need a decision here.

> For the immediate issue and what you've asked for above, I won't be able to do it unless the blank rows are removed from the target sheets. Also, I'd need to know if the column names you've used are the same. If they're not, it's quite a lot of work to write a custom translation for every sheet. 

Spaces are now removed. Column names are now the same as before. Sorry, mistakes on our end.

> Another thing: the one you've linked as the one we're currently using is called v3, whereas the others are v4 but appear to read from a sheet called v2.1. If v2.1 is the most up to date, is it possible to just read from the v2.1 sheets? What is the import to other sheets doing for you? And can we just avoid that or replicate it on the API side to make things easier and possibly less error-prone on the google-sheets side of things? I'd guess large imports of data between sheets is where we end up with errors when we try to import them from google.

At the moment v2.1 needs lots of changes before it could be used to export, and those changes are linked to lots of other things we need to think through etc.

> Just wondering too, is the error of some records not loading still happening? I've not been explicitly checking, but recent emails seem to have been a more consistent number.

We think so yes. We're setting up more active monitoring now, and hoping these changes should help. Importantly, these changes are *we think* quick and easy. While, the ones for #1590 are torny and need careful thought.

**Comment by markmacgillivray** [comment]
@JosephMcArthur could you take a look at the journals sheet you linked above? It seems to have nothing but ISSNs in it, shouldn't there be more data?

**Comment by natalianorori** [comment]
> @JosephMcArthur could you take a look at the journals sheet you linked above? It seems to have nothing but ISSNs in it, shouldn't there be more data?

That should be linked now, I was working on the sheet were it was importing from but fixed it

**Comment by markmacgillivray** [comment]
OK ta! Yeah it's showing up now.

**Comment by markmacgillivray** [comment]
All records successfully imported into new dev system, now testing to see if changes made to handle multiple sources are working within the permissions algorithm.

**Comment by markmacgillivray** [comment]
OK, the new dev system has imported all records, I've checked the counts and they match what is in the 3 new sheets you linked. Note, the email that are sent each day are from the old dev system - I haven't bothered fixing it because the old system doesn't run permissions any more anyway. I'll just disable updates on the old system.

I've directly tested new dev permissions endpoint for a journal, a publisher, and an affiliation, and the results look right to me.

For the new system, I can give you URLs to trigger updates to each of those 3 new sheets any time you want. I haven't scheduled any auto update yet.

Is that everything we need to close this issue? The discussion of what we may want to do later with the API data or with changing column names to match output names doesn't matter for now, from the looks of it, and is just a nice-to-have for later anyway, I think.

**Comment by JosephMcArthur** [comment]
> For the new system, I can give you URLs to trigger updates to each of those 3 new sheets any time you want.

Whenever you get a moment would be lovely.

> I haven't scheduled any auto update yet.

If we could keep the daily update that would be good - just in case.

> Is that everything we need to close this issue? 

Can we keep the import emails, or, have some other way to check how many have been imported? 

Given the problem here is that the import isn't always working in unpredictable ways we'd like to monitor how many are actually imported to check the issue is actually solved.

**Comment by markmacgillivray** [comment]
@JosephMcArthur if you happen to be working today those new sheets appear to just have #REF in them now, at least one of them still does, whereas before today they all had rows of data. It's not urgent, but if you happen to have time to repopulate them then when I'm up tomorrow I can finish off some tests of daily imports on the new system.

**Comment by JosephMcArthur** [comment]
I think I've resolved this for `publishers` and `affiliations`. Haven't quite cracked it for journals.

@natalianorori if you need to update journals data between now and Wednesday let me know. We may just need to copy and paste it.

**Comment by markmacgillivray** [comment]
OK, I've also experimented with different cache headers on our requests to google and that seems to be helping with getting fresher responses. All 3 sheets are now updated on the new dev system. The journals one does appear to have 17 records that lack an ID now though, but maybe you expected that? They just have #N/A and #VALUE!  in various fields. Or perhaps they're an artefact of the export from google, for some reason. Either way, it doesn't cause the upload to fail, it just causes the index to have 17 extra records than it needs, but they're not going to match any proper queries anyway.
